### ISSUE TITLE
Indexer, collection, event and sublevel initialization

### ISSUE THREAD

**Issue by Happypig375** [fsharp/fslang-suggestions#969]
# Indexer, collection, event and sublevel initialization

I propose we allow indexers, collections, and events to be assigned during construction. Currently, only properties and fields are allowed. An indexer is a property too, why can't we assign to it during construction?

As an extension to https://github.com/fsharp/fslang-suggestions/issues/877, I propose we allow:
```fs
let p = Builder([2] = initializeToIndexer, collection @= [addToCollection1; addToCollection2], event @= [addToEvent1; addToEvent2], Property = (SubLevelProp1 = 2, SubLevelProp2 = 1))
```

The existing way of approaching this problem in F# is
```fs
let p = Builder()
p.[2] <- initializeToIndexer
p.collection.Add addToCollection1
p.collection.Add addToCollection2
p.event.Add addToEvent1
p.event.Add addToEvent2
p.Property.SubLevelProp1 <- 2
p.Property.SubLevelProp2 <- 1
```
## Pros and Cons 

The advantages of making this adjustment to F# are
1. Conciseness
2. Convenience - property initialization works with more things!

The disadvantages of making this adjustment to F# are that `@=` may have already been defined in user code. However, as with `nameof`, this can be solved by a library intrinsic.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): M

Related suggestions: https://github.com/fsharp/fslang-suggestions/issues/877

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by dsyme** [comment]
I agree with this in principle, though I'm not convinced by the `@=` syntax - it may be possible to simply use `=`

I will mark this as approved in principle and we can sort out the details in prototype and RFC

**Comment by Happypig375** [comment]
We need to differentiate between assignment and calling `Add` without assigning because it is assumed that an instance has been given by the constructor.

**Comment by dsyme** [comment]
> We need to differentiate between assignment and calling Add without assigning because it is assumed that an instance has been given by the constructor.

Yes, if the collection property is settable this is true.  The `@=` syntax does not feel intuitve however.  Will think it over.

Events are not settable so would not need `@=`

**Comment by charlesroddie** [comment]
Of the four suggestsions here, I would use collection (and it's seen a lot in C# code for good reason) but not the others.

**Comment by Happypig375** [comment]
Maybe we can even have a syntax that it equivalent to C#'s collection initialiser, such as `ResizeArray<int>([] = [1; 2; 3])`

For collection in a property: `Builder(Collection[] = [1; 2; 3])`

**Comment by xp44mm** [comment]
this issue is very similar to vb's with statement:

```vb
With MyLabel
    .Height = 2000
    .Width = 2000
    .Caption = "This is MyLabel"
End With
```

**Comment by Happypig375** [comment]
@xp44mm 
No. This issue is a direct counterpart of VB (and C#)'s object initializers:
```vb
Dim MyLabel = New Label With {.Height = 2000, .Width = 2000, .Caption = "This is MyLabel"}
```
```cs
var MyLabel = new Label { Height = 2000, Width = 2000, Caption = "This is MyLabel" }
```
and collection initializers:
```vb
Dim labels = New List(Of Label) From { New Label(), New Label(), New Label() }
```
```cs
var labels = new List<Label> { new Label(), new Label(), new Label() }

**Comment by xp44mm** [comment]
Giving a prefix such as a dot `.` on a component gives the IDE the ability to prompt for component names.
```f#
let p = Builder(
    .[2] = initializeToIndexer, 
    .collection @= [addToCollection1; addToCollection2], 
    .event @= [addToEvent1; addToEvent2], 
    .Property = (.SubLevelProp1 = 2, .SubLevelProp2 = 1)
)
```

**Comment by jl0pd** [comment]
Just for the record it's possible to use collection, indexer and event initalizers even now with extension properties:

```fsharp
open System.Collections.Generic

type Dictionary<'key, 'value> with
    member inline dict.With with set keysAndValues =
        for k, v in keysAndValues do
            dict[k] <- v

new Dictionary<string, string>(
    With = [
        "key", "value"
        "key2", "value2"
    ]
) |> printfn "dict: %A"

open System.Collections.ObjectModel
open System.Collections.Specialized

type ObservableCollection<'a> with
    member inline col.OnCollectionChanged with set handlers =
        for handler in handlers do
            col.CollectionChanged.AddHandler handler

    member inline col.Items with set items =
        for item in items do
            col.Add item

new ObservableCollection<int>(
    OnCollectionChanged = [
        NotifyCollectionChangedEventHandler(
            fun sender e -> printfn "Collection changed. Action: %A; NewItems: %A; OldItems: %A" e.Action e.NewItems e.OldItems
        )
    ],
    Items = [ 1; 2; 3 ]
) |> printfn "observableCollection: %A"
```
Which works as people may expect, but have overhead of creating list and requires each property to have extension counterparts. I'm still waiting for direct language support and recommend to avoid this *solution*

**Comment by dsyme** [comment]
@Happypig375 Did you compare and contrast with https://github.com/fsharp/fslang-suggestions/issues/290?  thanks

**Comment by Happypig375** [comment]
For sublevel initialization, we can also allow the syntax of #290 for a single sublevel initialization. However, for multiple properties under one outer property it quickly becomes repetitive to have to specify the outer name over and over again. So we should allow the syntax of this issue. Meanwhile the collection initialization of #290 conflicts with directly setting the property which this issue avoids using new syntax.

**Comment by glchapman** [comment]
With current F#, an operator inspired by Clojure's doto composes well with _.Property shorthand for mutable object initialization. The operator is:
```F#
let inline (&>) (target: 't) ([<InlineIfLambda>] action: 't -> 'r) =
    action target |> ignore
    target
```
Here's an example initialization built using it:
```F#
let initRoot() =
    RootClass(Name = "Root")
    // Set only one nested prop 
    // Note that setting properties requires set_Property method
    &> _.Nested.set_Num(17)
    // Set multiple nested props
    &> into _.Nested (_.set_Num(42) >&> _.set_Name("Nested"))
    &> into _.NumList (add {4;8;15;16;23;42})
    &> _.Click.Add(fun _ -> printfn "Clicked 1")
    &> _.set_Item("foo", 1)
```

A compilable gist sketching out these ideas: [MutObjectInit.fs](https://gist.github.com/glchapman/900e0f052c34d1da920bddf025fa99e3) .

The code in the gist has not been extensively tested; it really is only an example of something possibly useful we can do with the language as it is.

**Comment by samueleaton** [comment]
One good use case is with `System.Text.Json.Nodes.JsonObject` constructor.

### C# `JsonObject` Example

The [example in C#](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom#code-try-1) is:

```c#
var forecastObject = new JsonObject
    {
        ["Date"] = new DateTime(2019, 8, 1),
        ["Temperature"] = 25,
        ["Summary"] = "Hot",
        ["DatesAvailable"] = new JsonArray(
            new DateTime(2019, 8, 1), new DateTime(2019, 8, 2)),
        ["TemperatureRanges"] = new JsonObject
        {
            ["Cold"] = new JsonObject
            {
                ["High"] = 20,
                ["Low"] = -10
            }
        }
    };
```

In F# we could create an empty JsonObject and manually assign values (e.g. `myObject["my_key"] <- "myValue"`).

JsonObject has a [constructor which takes an IEnumerable](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.nodes.jsonobject.-ctor?view=net-8.0#system-text-json-nodes-jsonobject-ctor(system-collections-generic-ienumerable((system-collections-generic-keyvaluepair((system-string-system-text-json-nodes-jsonnode))))-system-nullable((system-text-json-nodes-jsonnodeoptions)))), but then we'd have to cast all elements to key value pairs:

```f#
let myObject =
      JsonObject(
          properties =
              [
                  KeyValuePair("my_property1", "my_value1")
                  KeyValuePair("my_property2", "my_value2")
                  KeyValuePair("my_property3", JsonObject())
              ]
      )
```

### Possible F# Design

If we took inspiration for how C# does it, then we could maybe initialize objects that implement `IDictionary<String,_>` or `IEnumerable<KeyValuePair<String,_>>`  by doing:

```f#
let myObject = JsonObject(["my_snake_prop"] = "my_snake_value", ["myCamelProp"] = "myCameValue")
```

**Comment by T-Gro** [comment]
If you have a lot of code like this, a module-specific (=isolated) operator makes the JSON object creation code read even better than the here presented C# version:

```fsharp
let (=>) k v = KeyValuePair(k,v)

let myObject =
      JsonObject(
          properties =
              [|
                  "my_property1" => "my_value1"
                  "my_property1" => "my_value1"
                  "my_property1" => JsonObject()
              |]
      )
```

https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAOgBUEAPbAgKQhiQIDkYATBCAWACh5l0tdCAwjGDAEAY2wBLWhAIBxZAgBOksZ04BiJDADuAQyWoARAGYTATgCMR9R1HYUACgC8APgCUKANYoAbimcUAGkEDAA1PWAAVwQABT1JJUcvABpfd1t7FDwMAHkAIwArcQdnThQKyupaAuKJR3LKpoq4JW4lKVYAxubelABtKB6+kaMcgH1W9uwMawDXFDGMcd9ImOthkd6lybbEDtmjecWJ1eiEDY4trZ2p/Zm5txRqpFqSxwyr68qoAF1NlDuIA==

**Comment by brianrourkeboll** [comment]
For collection initialization, I wonder whether reusing a limited subset of computation expression notation could be a viable approach.

That is, perhaps a value of a type having an `Add` method with the appropriate signature could be initialized by an expression enclosed in curly braces wherein implicit or explicit yields would be translated to calls to `Add`. Potentially, instances of `yield!` could likewise be translated to calls to `AddRange`.

For example, given:

```fsharp
type C () =
    member val Collection = ResizeArray<int> ()
```

you could write

```fsharp
let c = C (Collection = { yield 1; yield 2; yield 3 })
```

or

```fsharp
let c = C (Collection = { 1; 2; 3 })
```

both of which would be translated by the compiler to

```fsharp
let c = C ()
c.Collection.Add 1
c.Collection.Add 2
c.Collection.Add 3
```

From what I have seen in the parser and compiler, I think this _might_ be feasible to implement in a backwards-compatible way, since previously either the curly braces would have needed to be preceded by a computation expression builder value (or `seq`), or else the contents of the curly braces would need to form a valid record initialization expression.

There are two potential confounders that immediately come to mind.

The first is `seq`-less sequence expressions with ranges, which are historically valid but have produced a warning since https://github.com/fsharp/fslang-suggestions/issues/1033 was implemented:

```fsharp
type C () =
    member val Collection = Seq.empty<int> with get, set

let c = C (Collection = { 1..10 })
```

The second would be a new ambiguity that would arise if support for the spread operator `...` https://github.com/fsharp/fslang-suggestions/issues/1253 were implemented for both records and computation expressions. If we decided to support translating `yield!` to a call to `AddRange` in collection initializers, and if we also decided to support `...` as an equivalent to `yield!` in computation expressions and collection expressions, something like the following would be syntactically ambiguous:

```fsharp
C (P = { ...x })
```

In theory, the types of `P` and `x` could be used to disambiguate between a collection initialization and an assignment of a record spread expression, but I have not looked into how hard that would be to do or what the implications for type inference would be.
